### PR TITLE
serialize ipset operation with the same name

### PR DESCRIPTION
**Pull request by YangKeao**
Signed-off-by: Yang Keao <keao.yang@yahoo.com>

The problem is that: ipset is not isolated with network namespace until 3.12. So if multiple `FlushIPSet` requests with the same ipset name for different container arrives the daemon, and executed at the same time, the operation will fail:

```
-> create -> flush -> add -> rename -> return successfully
-> create -> flush ------------------------> add (Fail!)
```

I have though about two different solutions for it:

1. Add a pid to the name of ipset to simulate the isolation
2. Use a lock to serialize the ipset operations

I choose the second solution, because the first one needs more modification, and the pid is not a good identifier for namespace. However, the second choice will bring performance degradation for every users (and the correctness relies on the logic of NetworkChaos).

Do you have better ideas about the fix?

### CODE REVIEW

**Review by STRRL — CHANGES_REQUESTED** [review]
rest LGTM 

Actually, I prefer the first solution. I think we could the container id as a part of ipset name.

File: pkg/chaosdaemon/server.go [hunk 1]
```diff
@@ -67,6 +71,7 @@ func newDaemonServer(containerRuntime string) (*DaemonServer, error) {
 	}
 
 	return &DaemonServer{
+		IPSetLocker:              locker.New(),
```
[the remark this answers is not in the document]
Comment (YangKeao) [reply pkg/chaosdaemon/server.go]: Fixed.

File: pkg/chaosdaemon/ipset_server.go [hunk 2]
```diff
@@ -40,10 +40,22 @@ func (s *DaemonServer) FlushIPSets(ctx context.Context, req *pb.IPSetsRequest) (
...
+		// It's a bad solution, only for the users who don't want to upgrade
+		// their linux version to 3.12 :(
+		ipset := ipset
+		s.IPSetLocker.Lock(ipset.Name)
 		err := flushIPSet(ctx, req.EnterNS, pid, ipset)
 		if err != nil {
+			s.IPSetLocker.Unlock(ipset.Name)
 			return nil, err
 		}
+		s.IPSetLocker.Unlock(ipset.Name)
```
[the remark this answers is not in the document]
Comment (YangKeao) [reply pkg/chaosdaemon/ipset_server.go]: Fixed

**Review by STRRL — APPROVED** [review]
LGTM

**Review by cwen0 — APPROVED** [review]
LGTM

**Pull request merged**